Attempt to fix #2: Hard parser crashes

seijikun opened the pull request:
- I replaced all uses of unwrap() in the parser and replaced them with error handling where possible.
Unfortunately, that required to rewrite some nice iterators to loops.
- Switched implementation of device discovery to use tokio sockets, which allowed to remove the `sleep(500ms)` hack, some unsafe code, and the `socket2` crate dependency. This further also seems to fix some problems with indefinite hangs I had - probably due to the mixture of tokio async and blocking code.
- Silently ignore parsing errors in device discovery stream, don't crash

**Warning:** Changed `discover_pnp_locations()` signature

I ran rustfmt on the commit, I hope that was ok!

Hopefully fixes #2

seijikun committed:
Remove uses of unwrap() in parser and replace by error handling

seijikun committed:
Device discovery: fix crashes for invalid response and simplify implementation

- Switched implementation of device discovery to use tokio sockets
- Silently ignore parsing errors during device discovery, don't crash
- Changed discover_pnp_locations() signature

seijikun commented:
I don't know if that fixes everything yet, because the problematic device isn't responding anymore - was probably switched off (wasn't mine).

tsirysndr commented:
Great work on the pull request! It's always good to see code improvements that remove unnecessary unsafe code and make the code more reliable.

Regarding the changes you made, it's great that you replaced the uses of  `unwrap()` with error handling where possible. 

However, as you've mentioned, you have made some breaking changes to the `discover_pnp_locations() ` function which has resulted in a change in its signature.

As a result, I would suggest updating the examples in the `README` file to reflect these changes. This will help new users understand how to use the updated function correctly.

Overall, the changes look good and should help improve the reliability and maintainability of the code. Thank you for your contributions!

seijikun committed:
Fix examples with new discover_pnp_locations() signature

seijikun committed:
Apply clippy lints

seijikun commented:
Yes! Please give me some more minutes, found some more I could clean up

seijikun committed:
Device_client: More error handling improvements, fix use of std::Mutex in async

- Using the std::sync::Mutex in async is dangerous, switched to tokio's Mutex
- Replaced more .unwrap() with error handling

seijikun commented:
Okay, I think now I'm done!

seijikun committed:
Fix examples
